### PR TITLE
feat(acir)!: Add a public outputs field

### PR DESCRIPTION
**Pull request by kevaundray**
# Related issue(s)

(If it does not already exist, first create a GitHub issue that describes the problem this Pull Request (PR) solves before creating the PR and link it here.)

Resolves (link to issue)

# Description

## Summary of changes

(Describe the changes in this PR. Point out breaking changes if any.)

## Dependency additions / changes

(If applicable.)

## Test additions / changes

(If applicable.)

# Checklist

- [ ] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` with default settings.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) this PR to the issue(s) that it resolves.
- [ ] I have reviewed the changes on GitHub, line by line.
- [ ] I have ensured all changes are covered in the description.

# Additional context

(If applicable.)

### CODE REVIEW

**Comment by TomAFrench** [comment]
Now we have an explicit return type on the ABI this should be able to be handled seamlessly by the evaluator. Would you like me to fix up any conflicts so we can get this merged?

**Comment by kevaundray** [comment]
@TomAFrench That would be great ser!

**Comment by kevaundray** [comment]
Just merged the conflicts -- One thing to double check is that the definition of public outputs here also matches the definition being used in Noir

File: acir/src/circuit/mod.rs [hunk 1]
```diff
@@ -22,6 +22,11 @@ pub struct Circuit {
     pub current_witness_index: u32,
     pub opcodes: Vec<Opcode>,
     pub public_inputs: PublicInputs,
+    // Public outputs are a subset of public inputs.
+    // They describe at a high level, what witness indices
+    // should be seen as the output of executing the
+    // function that this Circuit represents
+    pub public_outputs: PublicInputs,
```
Comment (TomAFrench) [inline acir/src/circuit/mod.rs:29]:
```suggestion
   // ACIR distinguishes between the public inputs which are provided
   // externally (`public_parameters`) or calculated within the circuit (`return_values`).
   // The elements of these sets may not be mutually exclusive, i.e. a parameter may be returned from the circuit.
   // All public inputs must be provided to the verifier at verification time.
   
    /// The set of public inputs provided by the prover.
    pub public_parameters: PublicInputs,
    /// The set of public inputs calculated within the circuit.
    pub return_values: PublicInputs,
```

Comment (TomAFrench) [reply acir/src/circuit/mod.rs:29]: We'd want to add a `fn public_inputs()-> PublicInputs` function which returns the union of these two sets as well.

**Review by TomAFrench — COMMENTED** [review]
Based on https://aztecprotocol.slack.com/archives/C0183F0V42V/p1676636370362909?thread_ts=1676636143.866669&cid=C0183F0V42V, we may want to take the opportunity to rename these fields for more clarity.

File: acir/src/circuit/mod.rs [hunk 2]
```diff
@@ -21,14 +21,27 @@ pub struct Circuit {
...
+    /// The set of public inputs provided by the prover.
+    pub public_parameters: PublicInputs,
+    /// The set of public inputs calculated within the circuit.
+    pub return_values: PublicInputs,
 }
 
 impl Circuit {
     pub fn num_vars(&self) -> u32 {
         self.current_witness_index + 1
     }
```
Comment (kevaundray) [inline acir/src/circuit/mod.rs:38]:
```suggestion
/// Returns all public inputs. This includes those provided as parameters to the circuit and those 
/// computed as return values. 
```

**Review by kevaundray — COMMENTED** [review]
LGTM - one small comment re docs

**Comment by kevaundray** [comment]
This LGTM -- @TomAFrench you'll need to approve since the PR was originally made by me

**Comment by TomAFrench** [comment]
Ah, that's true. Thanks. Approving now.

**Review by TomAFrench — APPROVED** [review]

**Pull request merged**